server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -520,114 +519,135 @@ public String toString() {
+                }
+                return builder.endObject();
+            }),
+
+            // nodes
+            chunkedSection(
+                metrics.contains(Metric.NODES),
+                (builder, params) -> builder.startObject("nodes"),
+                nodes.iterator(),
+                Iterators::single,
+                (builder, params) -> builder.endObject()
+            ),
+
+            // metadata
+            metrics.contains(Metric.METADATA) ? metadata.toXContentChunked(outerParams) : Collections.emptyIterator(),
+
+            // routing table
+            chunkedSection(
+                metrics.contains(Metric.ROUTING_TABLE),
+                (builder, params) -> builder.startObject("routing_table").startObject("indices"),
+                routingTable().iterator(),
+                indexRoutingTable -> Iterators.single((builder, params) -> {
+                    builder.startObject(indexRoutingTable.getIndex().getName());
+                    builder.startObject("shards");
+                    for (int shardId = 0; shardId < indexRoutingTable.size(); shardId++) {
+                        IndexShardRoutingTable indexShardRoutingTable = indexRoutingTable.shard(shardId);
+                        builder.startArray(Integer.toString(indexShardRoutingTable.shardId().id()));
+                        for (int copy = 0; copy < indexShardRoutingTable.size(); copy++) {
+                            indexShardRoutingTable.shard(copy).toXContent(builder, params);
+                        }
+                        builder.endArray();
+                    }
+                    return builder.endObject().endObject();
+                }),
+                (builder, params) -> builder.endObject().endObject()
+            ),
+
+            // routing nodes
+            chunkedSection(
+                metrics.contains(Metric.ROUTING_NODES),
+                (builder, params) -> builder.startObject("routing_nodes").startArray("unassigned"),
+                getRoutingNodes().unassigned().iterator(),
+                Iterators::single,
+                (builder, params) -> builder.endArray() // no endObject() here, continued in next chunkedSection()
+            ),
+            chunkedSection(
+                metrics.contains(Metric.ROUTING_NODES),
+                (builder, params) -> builder.startObject("nodes"),
+                getRoutingNodes().iterator(),
+                routingNode -> Iterators.concat(
+                    ChunkedToXContentHelper.startArray(routingNode.nodeId() == null ? "null" : routingNode.nodeId()),
+                    routingNode.iterator(),
+                    ChunkedToXContentHelper.endArray()
+                ),
+                (builder, params) -> builder.endObject().endObject()
+            ),
+
+            // customs
+            metrics.contains(Metric.CUSTOMS)
+                ? Iterators.flatMap(
+                    customs.entrySet().iterator(),
+                    cursor -> Iterators.concat(
```

Comment:
NIT: could use `org.elasticsearch.common.xcontent.ChunkedToXContentHelper#wrapWithObject` here to shorten this a little

Reply:
++ thanks, see c5bc46d577a